build/build/src/web.rs

```diff
@@ -0,0 +1,78 @@
+//! Module for managing the web parts of our codebase.
+//! This refers to JS/TS components, such as the Vue.js-based GUI, as well as the Electron client
+//! application.
+
+use crate::prelude::*;
+
+use ide_ci::programs::Npm;
+
+/// Result of root-level `npm install` call. Should not be accessed directly.
+static ONCE_INSTALL: tokio::sync::OnceCell<Result> = tokio::sync::OnceCell::const_new();
+
+/// This method invokes `npm install` in the root repository.
+///
+/// It ensures that at most one such invocation is running at any given time.
+/// If an invocation is already running, it waits for it to finish and then
+/// shares its result.
+///
+/// This method relies on the internal global state. It must not be called multiple times
+/// with different `repo_root` arguments.
+///
+/// It is useful, because otherwise the build script might end up invoking `npm install` multiple
+/// times, sometimes even in parallel. Unfortunately, in such cases, `npm` might fail with errors.
+pub fn install(repo_root: impl AsRef<Path>) -> BoxFuture<'static, Result> {
+    let root_path = repo_root.as_ref().to_owned();
+    async move {
+        let ret = ONCE_INSTALL
+            .get_or_init(async || Npm.cmd()?.with_current_dir(&root_path).install().run_ok().await)
+            .await;
+        ret.as_ref().copied().map_err(|e| anyhow!("Failed to install NPM dependencies: {e}"))
```

Comment:
This error mapping pattern looks like could be replaced with anyhow's [`context`](https://docs.rs/anyhow/latest/anyhow/index.html).

```suggestion
        ret.as_ref().copied().context("Failed to install NPM dependencies")
```

Reply:
Not in this case, as we have a reference to `Error` rather than the error itself.